corehq/apps/sms/views.py

```diff
@@ -665,36 +666,130 @@ def global_backend_map(request):
-                name = owner.raw_username
-            else:
-                url = reverse("case_details", args=[domain, owner._id])
-                if domain_obj.custom_case_username:
-                    name = owner.get_case_property(domain_obj.custom_case_username) or _("(unknown)")
-                else:
-                    name = owner.name
-            contacts.append({
-                "id" : owner._id,
-                "doc_type" : owner.doc_type,
-                "url" : url,
-                "name" : name,
-            })
     context = {
         "domain" : domain,
-        "contacts" : contacts,
     }
     return render(request, "sms/chat_contacts.html", context)
 
+
+def get_case_contact_info(domain_obj, case_ids):
+    data = {}
+    for doc in iter_docs(CommCareCase.get_db(), case_ids):
+        if domain_obj.custom_case_username:
+            name = doc.get(domain_obj.custom_case_username, _('unknown'))
+        else:
+            name = doc.get('name', _('unknown'))
+        data[doc['_id']] = [name]
+    return data
+
+
+def get_mobile_worker_contact_info(domain_obj, user_ids):
+    data = {}
+    for doc in iter_docs(CommCareUser.get_db(), user_ids):
+        user = CommCareUser.wrap(doc)
+        data[user.get_id] = [user.raw_username]
+    return data
+
+
+def get_contact_info(domain):
+    # If the data has been cached, just retrieve it from there
+    cache_key = 'sms-chat-contact-list-%s' % domain
+    cache_expiration = 30 * 60
+    try:
+        client = cache_core.get_redis_client()
+        cached_data = client.get(cache_key)
+        if cached_data:
+            return json.loads(cached_data)
+    except:
+        pass
+
+    verified_number_ids = VerifiedNumber.by_domain(domain, ids_only=True)
+    domain_obj = Domain.get_by_name(domain, strict=True)
+    case_ids = []
+    mobile_worker_ids = []
+    data = []
+    for doc in iter_docs(VerifiedNumber.get_db(), verified_number_ids):
+        owner_id = doc['owner_id']
+        if doc['owner_doc_type'] == 'CommCareCase':
+            case_ids.append(owner_id)
+            data.append([
+                None,
+                'case',
+                doc['phone_number'],
+                owner_id,
+            ])
+        elif doc['owner_doc_type'] == 'CommCareUser':
+            mobile_worker_ids.append(owner_id)
+            data.append([
+                None,
+                'mobile_worker',
+                doc['phone_number'],
+                owner_id,
+            ])
+    contact_data = get_case_contact_info(domain_obj, case_ids)
+    contact_data.update(get_mobile_worker_contact_info(domain_obj, mobile_worker_ids))
+    for row in data:
+        contact_info = contact_data.get(row[3])
+        row[0] = contact_info[0] if contact_info else _('(unknown)')
+
+    # Save the data to the cache for faster lookup next time
+    try:
+        client = cache_core.get_redis_client()
```

Comment:
probably not - updated (it might make a difference if it fails the first time, but it's probably not likely to succeed the second time if it fails the first time)